src/main/java/com/charlesgadeken/entwined/EngineController.java

```diff
@@ -0,0 +1,146 @@
+package com.charlesgadeken.entwined;
+
+import com.charlesgadeken.entwined.effects.TSEffectController;
+import com.charlesgadeken.entwined.effects.original.ScrambleEffect;
+import com.charlesgadeken.entwined.effects.original.SpeedEffect;
+import com.charlesgadeken.entwined.effects.original.SpinEffect;
+import heronarts.lx.LX;
+import heronarts.lx.effect.BlurEffect;
+import heronarts.lx.mixer.LXAbstractChannel;
+import heronarts.lx.mixer.LXChannel;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class EngineController {
+    LX lx;
+
+    public int baseChannelIndex;
+    public int numChannels;
+
+    public int startEffectIndex;
+    public int endEffectIndex;
+
+    public boolean isAutoplaying;
+    EntwinedAutomationRecorder automation;
+    boolean[] previousChannelIsOn;
+
+    public List<TSEffectController> effectControllers = new ArrayList<>();
+    public int activeEffectControllerIndex = -1;
+
+    public SpeedEffect speedEffect;
+    public SpinEffect spinEffect;
+    public BlurEffect blurEffect;
+    public ScrambleEffect scrambleEffect;
+
+    public EngineController(LX lx) {
+        this.lx = lx;
+    }
+
+    public List<LXChannel> getChannels() {
+        // NOTE(meawoppl) dangercast here. @Slee?
+        return lx.engine.mixer.getChannels()
+                .subList(baseChannelIndex, baseChannelIndex + numChannels).stream()
+                .map(c -> (LXChannel) c)
+                .collect(Collectors.toList());
+    }
+
+    public void setChannelPattern(int channelIndex, int patternIndex) {
+        // NOTE(meawoppl) - Needs review from @Slee?
+        // Previously this was:
+        //        if (patternIndex == -1) {
+        //            patternIndex = 0;
+        //        } else {
+        //            patternIndex++;
+        //        }
+        //        lx.engine.getChannel(channelIndex).goIndex(patternIndex);
+        // Trying....
+        lx.engine.mixer.setFocusedChannel(lx.engine.mixer.getChannel(channelIndex));
```

Comment:
This is a totally different functionality... the "focused channel" is setting which channel has "focus" in the UI - at any given time one channel has focus.

Not sure why it was changed what was there before? That conditional with the -1 doesn't make much sense (the ++ on the else would do the same thing?)

to preserve original functionality I think you could just do the following:

```java
lx.engine.mixer.getChannel(channelIndex).goIndex(patternIndex + 1);
```

Though I would look into why you are passing patternIndex values that are one too low...

Reply:
`LXAbstractChannel` doesn't seem to have a `.goIndex()` method?

Reply:
You're right - see other comment for explanation - you'll want to confirm here that the channel being selected is definitely a basic LXChannel with patterns on it.

Reply:
For reference, this is what an LXGroup looks like - lets you fade up/down, turn on/off etc. a group of channels together. Hopefully that makes it more clear what's going on here. The mixer can contain both these group channels as well as simple basic channels that are not part of a group.

<img width="474" alt="Screen Shot 2020-11-25 at 1 41 03 PM" src="https://user-images.githubusercontent.com/597850/100284349-e9ee4700-2f23-11eb-9d32-17a19dd91464.png">

Reply:
:thinking: Do you expect instances of `LXChannel` to have `goIndex()`? I don't find it there either...

Sorry if I am being a slow kid here...

Reply:
Ack my apologies, that was renamed for clarity (and porting confusion) to `goPatternIndex` - thought that "goIndex" was way too vague... there is also FYI a `goPattern` method if you're passing a handle to the `LXPattern` itself

Reply:
`goPatternIndex()` :facepalm: